### DISCUSSION TITLE
Serverless / Decentralized Appwrite; Specifically, a low-latency Primary Database Cluster or Blockchain serving distinct Appwrite installations and Frontends

### DISCUSSION THREAD

**Discussion by zdanl** [appwrite/appwrite#5076] [Ideas]
![download](https://user-images.githubusercontent.com/114028070/216675509-a0020bc2-674b-4e00-8262-49fb98d02d3f.png)
![download](https://user-images.githubusercontent.com/114028070/216675573-418db477-73a5-46d3-861a-6d993a6ace86.jpg)

Probably going to be ignored / moved, but in the context of Appwrite Cloud and the future prospect of saving even more time & cost developing ever **increasingly complex** Web & Mobile Apps (with Appwrite), I would like to speak some about Decentralized / Serverless Backend approaches, not just Self-Hosted and Open Source. And about flexibilty, perhaps, thereby.

This found the way into my attention, which I think you should know about. It promises more than it serves, currently, but the idea is nicely visualized. For performance reasons, IPFS clearly is not a solution, and the Chain isn't as free/gasless as it proclaims.

https://www.gelato.network/

Where i am going here, is turning Appwrite BaaS into a sort of a ghost protocol defining opionated APIs (Auth, Storage, Functions, maybe Payment Providers, Crypto and web3 in the SDK in the future) and turning all projects based on Appwrite into Backendless dApps. No more need for a centralized, but self-hosted, MariaDB and PHP7 Backend. In theory. But EVM Smartcontracts / AssemblyScript to WASM WebAssembly, Serverless functions, executed on-chain, which is being worked on.

Leaves out the problem of Storage / Database.

Redis Cluster Cloud https://redis.io/docs/management/scaling/ callable from Browser Javascript and NodeJS by API as a Low-Latency Encrypted Primary Database with AES 256 CBC queryable blobs, JSON documents and collections abstracted, exist;  Encryption per User or Team rendering Permissionless, can be done. Hardware Accelerated Crypto Functions exist, harnessing things like Intel/AMD CPU Extensions or even GPU.js

Essentially, for the low price of $7/h, there is an official Redis Cluster Cloud solution offerating an insane amount of shards, multi-region high-availability and infinite database size and resources, with SSD and RAM upgrade possible. 
The cost can be substantially lower at 0.33$/h per hour without all that. 

There is a scientific paper on the theory of why public databases containing confidential records are secure if those are encrypted, although visible to everybody. In my point of view, even with granularity, currently rendered by server side logics due to the Permissions controller. 
Ethereum On-Chain Data Storage of up to 24 KB per transaction is not ... feasible due to cost, and probably latency.

![download](https://user-images.githubusercontent.com/114028070/216671799-72e4a966-136e-4e58-9202-5e3fa69b3ece.png)
 
![download](https://user-images.githubusercontent.com/114028070/216675424-655f3bd4-1e41-48ef-b5e1-bd1093529d4a.jpg)

in short: I am bouncing the idea of a Serverless Appwrite and was looking for a low-latency decentralized / at least high-availability , storage solution, that can be accessed from the Browser, such as Serverless functions (or EVM Blockchain Smartcontracts over the Infura / Moralis API) can . Or Vercel. Aprvpwrite Functions are *blazing fast* in comparison to Serverless, so I am hesitant. 

Teams as a form of Roles/Privileges/Permissions extended to the Appwrite.client could be rendered as Bitfields and Enums instead, for performance. Bitfields compiled to Integer after binary AND |.

https://github.com/aesy/easy-bits

</braindump end>

**Comment by zdanl** [comment]
https://eprint.iacr.org/2020/827.pdf
